File: mirea/cources/2026-2027/aii-july/04/contest/B.cpp
// Перед ответом на запросы для каждого префикса и каждого суффикса
// заранее посчитаем количество уникальных чисел. Для этого идём вдоль массива,
// вставляем очередной элемент в set и сохраняем размер set после вставки.
// При ответе на запросы берём из массива заранее посчитанные значения и выводим их.
#include <bits/stdc++.h>
using namespace std;
void solve() {
    // читаем массив
    int n, q;
    cin >> n >> q;
    vector<int> a(n);
    for (int i = 0; i < n; i++)
        cin >> a[i];
    // считаем и запоминаем количество уникальных элементов на каждом префиксе:
    set<int> numbers;
    vector<int> p(n);
    for (int i = 0; i < n; i++) {
        numbers.insert(a[i]);
        p[i] = numbers.size();
    }
    // аналогично для каждого суффикса:
    numbers.clear();
    vector<int> s(n);
    for (int i = n-1; i >= 0; i--) {
        numbers.insert(a[i]);
        s[i] = numbers.size();
    }
    // отвечаем на запросы:
    while (q --> 0) {
        char c; int i;
        cin >> c >> i;
        i--;
        cout << (c == 'p' ? p[i] : s[i]) << '\n';
    }
}
main() {
    int tt; cin >> tt;
    while (tt --> 0) solve();
}

File: mirea/cources/2026-2027/aii-july/04/contest/B.py
# Перед ответом на запросы для каждого префикса и каждого суффикса
# заранее посчитаем количество уникальных чисел. Для этого идём вдоль массива,
# вставляем очередной элемент в set и сохраняем размер set после вставки.
# При ответе на запросы берём из массива заранее посчитанные значения и выводим их.
def solve():
    # читаем массив
    n, q = map(int, input().split())
    a = list(map(int, input().split()))
    # считаем и запоминаем количество уникальных элементов на каждом префиксе:
    numbers = set()
    p = [0] * n
    for i in range(n):
        numbers.add(a[i])
        p[i] = len(numbers)
    # аналогично для каждого суффикса:
    numbers.clear()
    s = [0] * n
    for i in range(n - 1, -1, -1):
        numbers.add(a[i])
        s[i] = len(numbers)
    # отвечаем на запросы:
    while q > 0:
        c, i = input().split()
        i = int(i) - 1
        print(p[i] if c == 'p' else s[i])
        q -= 1


tt = int(input())
while tt > 0:
    solve()
    tt -= 1